packages/edit-site/src/components/sidebar-edit-mode/page-panels/hooks.js

```diff
@@ -86,3 +92,86 @@ export function useCurrentTemplateSlug() {
 	return templates?.find( ( template ) => template.slug === entityTemplate )
 		?.slug;
 }
+
+// This is duplicated.
+const filterOutDuplicatesByName = ( currentItem, index, items ) =>
+	index === items.findIndex( ( item ) => currentItem.name === item.name );
+
+function injectThemeAttributeInBlockTemplateContent(
+	block,
+	currentThemeStylesheet
+) {
+	block.innerBlocks = block.innerBlocks.map( ( innerBlock ) => {
+		return injectThemeAttributeInBlockTemplateContent(
+			innerBlock,
+			currentThemeStylesheet
+		);
+	} );
+
+	if (
+		block.name === 'core/template-part' &&
+		block.attributes.theme === undefined
+	) {
+		block.attributes.theme = currentThemeStylesheet;
+	}
+	return block;
+}
+
+function preparePatterns( patterns, template, currentThemeStylesheet ) {
+	return (
+		patterns
+			.filter(
+				( pattern ) => ! PATTERN_CORE_SOURCES.includes( pattern.source )
+			)
+			.filter( filterOutDuplicatesByName )
+			// Filter only the patterns that are compatible with the current template.
+			.filter( ( pattern ) =>
+				pattern.templateTypes?.includes( template.slug )
+			)
+			.map( ( pattern ) => ( {
+				...pattern,
+				keywords: pattern.keywords || [],
+				type: PATTERN_TYPES.theme,
+				blocks: parse( pattern.content, {
+					__unstableSkipMigrationLogs: true,
+				} ).map( ( block ) =>
+					injectThemeAttributeInBlockTemplateContent(
+						block,
+						currentThemeStylesheet
+					)
+				),
+			} ) )
+	);
+}
+
+// Should we also get templates?
+export function useAvailablePatterns( template ) {
+	const currentThemeStylesheet = useSelect(
+		( select ) => select( coreStore ).getCurrentTheme().stylesheet
+	);
+
+	let availablePatterns = useSelect( ( select ) => {
+		const { getSettings } = unlock( select( editSiteStore ) );
+		const settings = getSettings();
+
+		const blockPatterns =
+			settings.__experimentalAdditionalBlockPatterns ??
+			settings.__experimentalBlockPatterns;
+
+		const restBlockPatterns = select( coreStore ).getBlockPatterns();
+
+		const patterns = [
+			...( blockPatterns || [] ),
+			...( restBlockPatterns || [] ),
+		];
+		return patterns;
+	} );
+
+	availablePatterns = preparePatterns(
+		availablePatterns,
+		template,
+		currentThemeStylesheet
+	);
+
+	return availablePatterns;
```

Comment:
Suggestion: The `useAvailablePatterns` could use some improvements:

* Don't merge patterns inside the `useSelect` to avoid returning new values on each call.
* Let's combine selectors into one. Both are subscribed to a core store; keeping them separate doesn't improve anything.
* Memoize `availablePatterns`. Parsing blocks can be expensive, and `useAsyncList` expects stable references.